File: backend/crates/storage/src/dto/athlete.rs
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;
use uuid::Uuid;
use validator::Validate;

/// Response containing basic athlete information
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct AthleteResponse {
    pub athlete_id: Uuid,
    pub first_name: String,
    pub last_name: String,
    pub slug: String,
    pub gender: String,
    pub nationality: Option<String>,
    pub country: String,
    pub profile_picture_url: Option<String>,
    pub created_at: NaiveDateTime,
}

/// Detailed athlete response with competition history
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct AthleteDetailResponse {
    pub athlete_id: Uuid,
    pub first_name: String,
    pub last_name: String,
    pub slug: String,
    pub gender: String,
    pub nationality: Option<String>,
    pub country: String,
    pub profile_picture_url: Option<String>,
    pub created_at: NaiveDateTime,
    pub competitions: Vec<AthleteCompetitionSummary>,
    pub personal_records: Vec<PersonalRecord>,
    pub total_competitions: i64,
}

/// Summary of athlete's performance in a competition
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct AthleteCompetitionSummary {
    pub competition_id: Uuid,
    pub competition_name: String,
    pub competition_slug: String,
    pub competition_date: Option<chrono::NaiveDate>,
    pub category_name: String,
    pub rank: Option<i32>,
    pub total: rust_decimal::Decimal,
    pub ris_score: Option<rust_decimal::Decimal>,
    pub is_disqualified: bool,
}

/// Personal record for a specific movement
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct PersonalRecord {
    pub movement_name: String,
    pub max_weight: rust_decimal::Decimal,
    pub competition_name: String,
    pub competition_slug: String,
    pub date: Option<chrono::NaiveDate>,
}

/// Request payload for creating a new athlete
#[derive(Debug, Clone, Serialize, Deserialize, Validate, ToSchema)]
pub struct CreateAthleteRequest {
    #[validate(length(
        min = 1,
        max = 255,
        message = "First name must be between 1 and 255 characters"
    ))]
    pub first_name: String,

    #[validate(length(
        min = 1,
        max = 255,
        message = "Last name must be between 1 and 255 characters"
    ))]
    pub last_name: String,

    #[validate(custom(function = "validate_gender"))]
    pub gender: String,

    #[validate(length(max = 255))]
    pub nationality: Option<String>,

    #[validate(length(min = 1, max = 255, message = "Country is required"))]
    pub country: String,

    #[validate(url)]
    #[validate(length(max = 500))]
    pub profile_picture_url: Option<String>,
}

/// Request payload for updating an existing athlete
#[derive(Debug, Clone, Serialize, Deserialize, Validate, ToSchema)]
pub struct UpdateAthleteRequest {
    #[validate(length(min = 1, max = 255))]
    pub first_name: Option<String>,

    #[validate(length(min = 1, max = 255))]
    pub last_name: Option<String>,

    #[validate(custom(function = "validate_gender"))]
    pub gender: Option<String>,

    #[validate(length(max = 255))]
    pub nationality: Option<String>,

    #[validate(length(min = 1, max = 255))]
    pub country: Option<String>,

    #[validate(url)]
    #[validate(length(max = 500))]
    pub profile_picture_url: Option<String>,
}

// Validation helper
fn validate_gender(gender: &str) -> Result<(), validator::ValidationError> {
    const VALID_GENDERS: &[&str] = &["M", "F", "MX"];

    if VALID_GENDERS.contains(&gender) {
        Ok(())
    } else {
        Err(validator::ValidationError::new("invalid_gender"))
    }
}

impl From<crate::models::Athlete> for AthleteResponse {
    fn from(athlete: crate::models::Athlete) -> Self {
        Self {
            athlete_id: athlete.athlete_id,
            first_name: athlete.first_name,
            last_name: athlete.last_name,
            slug: athlete.slug,
            gender: athlete.gender,
            nationality: athlete.nationality,
            country: athlete.country,
            profile_picture_url: athlete.profile_picture_url,
            created_at: athlete.created_at,
        }
    }
}

File: frontend/src/lib/types/athlete.ts
import type { Attempt, RisSource } from './competition';

export interface AthleteSummary {
  athlete_id: string;
  first_name: string;
  last_name: string;
  slug: string;
  gender: string;
  country: string;
  profile_picture_url: string | null;
  created_at: string;
}

export interface StandingPlace {
  place: number;
  field: number;
}

export interface CountryStanding extends StandingPlace {
  code: string;
}

export interface RisStanding {
  score?: string | null;
  global: StandingPlace;
  country: CountryStanding;
}

export interface WeightClassStanding {
  class: string;
  total?: string | null;
  global: StandingPlace;
  country: CountryStanding;
}

export interface MetricStanding {
  value: string;
  class: string;
  global: StandingPlace;
  country: CountryStanding;
}

export interface AthleteStanding {
  ris?: RisStanding | null;
  weight_class?: WeightClassStanding | null;
  total?: MetricStanding | null;
  muscleup?: MetricStanding | null;
  pullup?: MetricStanding | null;
  dips?: MetricStanding | null;
  squat?: MetricStanding | null;
}

export interface AthleteDetail {
  athlete_id: string;
  first_name: string;
  last_name: string;
  native_name?: string | null;
  slug: string;
  gender: string;
  country: string;
  profile_picture_url: string | null;
  instagram_handle: string | null;
  created_at: string;
  competitions: AthleteCompetitionSummary[];
  personal_records: PersonalRecord[];
  total_competitions: number;
  standing?: AthleteStanding | null;
}

export interface AthleteLift {
  movement_name: string;
  best_weight: string | null;
  attempts: Attempt[];
}

export interface AthleteCompetitionSummary {
  competition_id: string;
  competition_name: string;
  competition_slug: string;
  competition_date: string | null;
  category_name: string;
  division?: string | null;
  rank: number | null;
  total: string | null;
  ris_score: string | null;
  ris_source: RisSource | null;
  status: 'competed' | 'disqualified' | 'no_show';
  event?: string | null;
  lifts: AthleteLift[];
}

export interface PersonalRecord {
  movement_name: string;
  max_weight: string;
  competition_name: string;
  competition_slug: string;
  date: string | null;
}
